stdlib/public/core/Array.swift

```diff
@@ -1865,3 +1865,129 @@ internal struct _ArrayAnyHashableBox<Element: Hashable>
     return true
   }
 }
+
+extension Array : Differentiable where Element : AdditiveArithmetic & Differentiable {
+  public struct TangentVector : AdditiveArithmetic & Differentiable {
+    public typealias TangentVector = Array<Element>.TangentVector
+    public typealias CotangentVector = Array<Element>.CotangentVector
+
+    public var elements: [Element.TangentVector]
+
+    public static var zero: TangentVector {
+      get {
+        return TangentVector(elements: [])
+      }
+    }
+    
+    public init(elements: [Element.TangentVector]) {
+      self.elements = elements
+    }
+
+    public func count() -> Int {
+      return elements.count
+    }
+
+    public static func + (lhs: TangentVector, rhs: TangentVector) -> TangentVector {
+      return TangentVector(elements: zipLongest(lhs.elements, rhs.elements).map {
+        ($0 ?? .zero) + ($1 ?? .zero)
+      })
+    }
+
+    public static func - (lhs: TangentVector, rhs: TangentVector) -> TangentVector {
+      return TangentVector(elements: zipLongest(lhs.elements, rhs.elements).map {
+        ($0 ?? .zero) - ($1 ?? .zero)
+      })
+    }
+
+    @inlinable @inline(__always)
+    public func moved(along direction: TangentVector) -> [Element.TangentVector] {
+      return zip(elements, direction.elements).map { $0.moved(along: $1) }
+    }
+
+    @inlinable @inline(__always)
+    public func tangentVector(from cotangent: CotangentVector) -> TangentVector {
+      let e = zip(elements, cotangent.elements).map { $0.tangentVector(from: $1) }
+      return TangentVector(elements: e)
+    }
+  }
+
+  public struct CotangentVector : AdditiveArithmetic & Differentiable {
+    public typealias TangentVector = Array<Element>.CotangentVector
+    public typealias CotangentVector = Array<Element>.TangentVector
+
+    public var elements: [Element.CotangentVector]
+
+    public static var zero: TangentVector {
+      get {
+        return TangentVector(elements: [])
+      }
+    }
+    
+    public init(elements: [Element.CotangentVector]) {
+      self.elements = elements
+    }
+
+    public func count() -> Int {
+      return elements.count
+    }
+
+    public static func + (lhs: TangentVector, rhs: TangentVector) -> TangentVector {
+      return TangentVector(elements: zipLongest(lhs.elements, rhs.elements).map {
+        ($0 ?? .zero) + ($1 ?? .zero)
+      })
+    }
+
+    public static func - (lhs: TangentVector, rhs: TangentVector) -> TangentVector {
+      return TangentVector(elements: zipLongest(lhs.elements, rhs.elements).map {
+        ($0 ?? .zero) - ($1 ?? .zero)
+      })
+    }
+
+    @inlinable @inline(__always)
+    public func moved(along direction: TangentVector) -> [Element.CotangentVector] {
```

Comment:
The direction type should be the parent type's `TangentVector`. Self's tangent vector is equal to self, per `CotangentVector.TangentVector == CotangentVector` axiom in the protocol.

```suggestion
    public func moved(along direction: Array.TangentVector) -> [Element.CotangentVector] {
```